---
title: "RNAseq"
output: html_document
---


# Background


# load biomart annotations

```{r eval=FALSE}
library(biomaRt)
biomaRt::listEnsemblArchives()
listMarts()

mart_aug2020 <- biomaRt::useMart(biomart = "ENSEMBL_MART_ENSEMBL", 
                         dataset = "hsapiens_gene_ensembl",
                         host = "http://aug2020.archive.ensembl.org")

head(listAttributes(mart_aug2020), n = 100)
grep( "entrez", listAttributes(mart_aug2020)$name,value = T)

t2g_aug2020 <- biomaRt::getBM(attributes = c("entrezgene_id","external_gene_name", "ensembl_gene_id", "ensembl_gene_id_version", "gene_biotype", "transcript_biotype","chromosome_name", "band", "transcript_length", "start_position", "end_position", "strand"), mart = mart_aug2020)
t2g_aug2020 <- dplyr::rename(t2g_aug2020,  entrez_gene = entrezgene_id, ext_gene = external_gene_name, ens_gene = ensembl_gene_id, ens_gene_ver = ensembl_gene_id_version)

save(list = c("t2g_aug2020", "mart_aug2020"), file = "/researchers/antonio.ahn/general/R/Download_resources/RData/mart_aug2020.RData")
```


```{r}
load("/researchers/antonio.ahn/general/R/Download_resources/RData/mart_aug2020.RData")

# apr2019 is the last version to not have converged the lncRNA biotypes into just one group
# Starting from July2019, lncRNA biotypes are converged to a single "lncRNA" category.... why?? 

load("/researchers/antonio.ahn/general/R/Download_resources/RData/mart_apr2019.RData")
```

# load data 

```{r echo=FALSE, results=FALSE, message=FALSE}
library(tidyverse)
library(ggrepel)
library(ggpubr)
library(cowplot)
library(DESeq2)
library(DT)
library(ggplot2)
library(RColorBrewer)
library(biomaRt)
library(ggsci)
library(pheatmap)
library(gplots)
library(ggdendro)
library(edgeR)
library(fgsea)
library("IHW")
```

```{r results=FALSE, message=FALSE}
fcounts <- read_tsv(file = "/researchers/krutika.ambani/Goel_lab_members/Keefe_Chan/220714_sgRB1_sgTP53_sgControl_DMSO_DOX_LY/results/bamfiles/featurecounts_stranded_2/featurecounts_stranded_2.txt", skip = 1)
```

```{r echo=FALSE, results=FALSE}
# separating out the gene name annotations to the gene expression matrix
# gene name annotations
GTF_annotations <- dplyr::select(fcounts, Geneid:Length)
# gene expression matrix
fcounts_mat <- dplyr::select(fcounts, -c(Chr:Length)) %>% data.frame(row.names = 1)
fcounts_mat %>% colnames

new_colnames <- gsub("Aligned.sortedByCoord.out.bam","",
     gsub("X.researchers.krutika.ambani.Goel_lab_members.Keefe_Chan.220714_sgRB1_sgTP53_sgControl_DMSO_DOX_LY.results.bamfiles.","",colnames(fcounts_mat)))

# remove the S numbers for example "_S2"
new_colnames <- new_colnames %>% gsub("_H.*", "", .)

# changing the clone names and changing rep names so its easier to separate later
new_colnames <- new_colnames %>% gsub("\\.", "_", .) %>% 
  gsub("_2_6", "_2-6", .) %>% 
  gsub("_4_1", "_4-1", .) %>% 
  gsub("_1_3", "_1-3", .) %>% 
  gsub("_4_2", "_4-2", .) %>% 
  gsub("Rep_1", "Rep1", .) %>% 
  gsub("Rep_2", "Rep2", .) %>% 
  gsub("Rep_3", "Rep3", .) 


# change from old to new names
colnames(fcounts_mat) <- new_colnames

# create sample information
sample_information <- data.frame(name = colnames(fcounts_mat), samples = colnames(fcounts_mat)) %>% separate(samples, into = c("knockout","clone","treatment","replicate"), sep = "_")
sample_information$treatment <- factor(sample_information$treatment, levels = c("DMSO","Abemaciclib","Doxorubicin"), labels = c("DMSO","LY","DOX"))
sample_information$replicate <- factor(sample_information$replicate)

```


```{r eval=FALSE, echo=FALSE}
# how many genes are expressed or not expressed in each group?

df1 <- table(fcounts_mat[,1:3] %>% rowSums(fcounts_mat > 5) >= 3) %>% data.frame %>% mutate(group = "DMSO")
df2 <- table(fcounts_mat[,4:6] %>% rowSums(fcounts_mat > 5) >= 3) %>% data.frame %>% mutate(group = "DOX")
df3 <- table(fcounts_mat[,7:9] %>% rowSums(fcounts_mat > 5) >= 3) %>% data.frame %>% mutate(group = "LY")
df4 <- table(fcounts_mat[,10:12] %>% rowSums(fcounts_mat > 5) >= 3) %>% data.frame %>% mutate(group = "RAD")

expressed_df <- rbind(df1,df2,df3,df4)

expressed_df <- dplyr::rename(expressed_df, express = "Var1")
expressed_df$express <- expressed_df$express %>% factor(levels = c("FALSE", "TRUE"), labels = c("not_expressed", "expressed"))
expressed_df$group <- expressed_df$group %>% factor

expressed_df %>% ggplot(aes(x = group, y = Freq, fill = express)) + geom_bar(stat = "identity") + facet_grid(.~ express) + theme_bw()
```


# filtering

```{r results=FALSE}
# 60656 genes (rows) and 48 samples (columns)
dim(fcounts_mat)
# 15111 genes with zero counts in all 12 samples
table(rowSums(fcounts_mat == 0) == 48)

# Here i chose to keep genes that have higher or equal to count than 5 in at least 3 or more samples (there are 3 samples in my smallest comparison group)
index <- rowSums(fcounts_mat > 5) >= 3
#index
#FALSE  TRUE 
#38707 21949  
table(index)

# This reduces the number of genes
# filtered matrix = fmat 
fcounts_Fmat <- fcounts_mat[index,]

# no duplications in the ENSG ID names
rownames(fcounts_Fmat) %>% duplicated %>% table
rownames(fcounts_mat) %>% duplicated %>% table
```

```{r eval=FALSE}

write_tsv(fcounts_Fmat %>% as_tibble(
rownames = "ensembl_gene_id"), "/researchers/krutika.ambani/Goel_lab_members/Keefe_Chan/220714_sgRB1_sgTP53_sgControl_DMSO_DOX_LY/R_analysis/220714_intiial_QC_AA/1.QC_exploratory/data/220714_sgRB1_sgTP53_sgControl_DMSO_DOX_LY_fcounts_Fmat.txt")
```

```{r echo=FALSE, results=FALSE}
# making a count matrix with the gene symbols as rownames 
fcounts_mat %>% class

fcounts_mat_genesymbol <- fcounts_mat %>% as.matrix

table(rownames(fcounts_mat_genesymbol) %in% t2g_aug2020$ens_gene_ver)
# FALSE  TRUE 
#   44 60612
# i'm not sure what these 44 are. They look are from chrY
# "ENSG00000228572.7_PAR_Y"  "ENSG00000182378.14_PAR_Y" "ENSG00000178605.13_PAR_Y" "ENSG00000226179.6_PAR_Y" 
# filter(GTF_annotations, Geneid == "ENSG00000228572.7_PAR_Y")
rownames(fcounts_mat_genesymbol)[!rownames(fcounts_mat_genesymbol) %in% t2g_aug2020$ens_gene_ver]

rownames(fcounts_mat_genesymbol) <- t2g_aug2020$ext_gene[match(rownames(fcounts_mat_genesymbol), t2g_aug2020$ens_gene_ver)]

# the 44 that are not found are just NAs
rownames(fcounts_mat_genesymbol) %>% is.na %>% table
```

```{r results=FALSE, echo=FALSE}
# i think using this is better for making the single gene boxplots 
# because rlog shrinks lowly expressing genes, it is harder to interpret (although its better for calculating distances when generating PCA and Hclustering)
dds <- DESeqDataSetFromMatrix(fcounts_Fmat, colData = sample_information, design = ~ treatment)
dds <- DESeq(dds)

norm_mat <- counts(dds, normalized=TRUE)
norm_mat <- log2(norm_mat + 1)
rownames(norm_mat) <-  t2g_aug2020$ext_gene[match(rownames(norm_mat), t2g_aug2020$ens_gene_ver)]
```

```{r eval=FALSE, echo=FALSE}
saveRDS(norm_mat, file = "/researchers/krutika.ambani/Goel_lab_members/Keefe_Chan/220714_sgRB1_sgTP53_sgControl_DMSO_DOX_LY/R_analysis/220714_intiial_QC_AA/1.QC_exploratory/data/MCF7M_sgRB1_sgTP53_sgControl_RNAseq_norm_mat.rds")
```


# rlog normalisation (for Hclustering and PCA plots)

```{r results=FALSE, eval=FALSE}
dds <- DESeqDataSetFromMatrix(fcounts_Fmat, colData = sample_information, design = ~ treatment)
# adding rowData. This is not required... dunno why i did this previously
table(rownames(dds) %in% t2g_aug2020$ens_gene_ver)
rowData(dds) <- data.frame(ens_gene_ver = rownames(dds), SYMBOL = t2g_aug2020$ext_gene[match(rownames(dds), t2g_aug2020$ens_gene_ver)])

rld <- rlog(dds) %>% assay

save(list = c("dds", "rld"), file = "/researchers/krutika.ambani/Goel_lab_members/Keefe_Chan/220714_sgRB1_sgTP53_sgControl_DMSO_DOX_LY/R_analysis/220714_intiial_QC_AA/1.QC_exploratory/data/rld.RData")
```

```{r}
load("/researchers/krutika.ambani/Goel_lab_members/Keefe_Chan/220714_sgRB1_sgTP53_sgControl_DMSO_DOX_LY/R_analysis/220714_intiial_QC_AA/1.QC_exploratory/data/rld.RData")
# adding in the gene symbols instead of ENSG IDs
table(rownames(rld) %in% t2g_aug2020$ens_gene_ver)
rownames(rld) <- t2g_aug2020$ext_gene[match(rownames(rld), t2g_aug2020$ens_gene_ver)]
```


# QC

## non-filtered non-normalised
```{r echo=FALSE, results=FALSE}
fcounts_mat_ggplot <- log2(fcounts_mat+1) %>% t %>% data.frame
fcounts_mat_ggplot$samples <- rownames(fcounts_mat_ggplot)

fcounts_mat_ggplot$samples == sample_information$name
fcounts_mat_ggplot$group <- sample_information$treatment

fcounts_mat_ggplot <- tidyr::gather(fcounts_mat_ggplot, key = genes, value = values, -samples, -group)

fcounts_mat_ggplot$samples <- factor(fcounts_mat_ggplot$samples, levels = unique(fcounts_mat_ggplot$samples))

# filtered for genes that are not expressed or lowly expressed in all 10 samples

a1 <- fcounts_mat_ggplot %>% ggplot(aes(x=values, group = samples, col = group)) + 
  geom_density() +
  ggtitle("counts (log2) (before rlog normalisation)") +
#  scale_color_manual(values = redblue_col) +
  xlab("counts (log2)") +
  theme_bw() +
  theme(axis.title = element_text(size = 12),
        axis.text = element_text(size = 10))

a2 <- fcounts_mat_ggplot %>% ggplot(aes(y=values, x = samples, col = group)) + 
  geom_boxplot() +
    ggtitle("counts (log2) (before rlog normalisation and non filtered)") +
#  scale_color_manual(values = redblue_col) +
  ylab("counts (log2)") +
  theme_bw() +
  theme(axis.title = element_text(size = 12),
        axis.title.x = element_blank(),
        axis.text = element_text(size = 10),
        axis.text.x = element_text(size = 5, angle = 45, hjust = 1))
```

```{r fig.width = 12}
plot_grid(a1,a2)
```

## filtered and non-normalised

```{r echo=FALSE, results=FALSE}
fcounts_Fmat_ggplot <- log2(fcounts_Fmat+1) %>% t %>% data.frame
fcounts_Fmat_ggplot$samples <- rownames(fcounts_Fmat_ggplot)

fcounts_Fmat_ggplot$samples == sample_information$name
fcounts_Fmat_ggplot$group <- sample_information$treatment

fcounts_Fmat_ggplot <- tidyr::gather(fcounts_Fmat_ggplot, key = genes, value = values, -samples, -group)

fcounts_Fmat_ggplot$samples <- factor(fcounts_Fmat_ggplot$samples, levels = unique(fcounts_Fmat_ggplot$samples))

# filtered for genes that are not expressed or lowly expressed in all 10 samples

a1 <- fcounts_Fmat_ggplot %>% ggplot(aes(x=values, group = samples, col = group)) + 
  geom_density() +
  ggtitle("counts (log2) (before rlog normalisation)") +
#  scale_color_manual(values = redblue_col) +
  xlab("counts (log2)") +
  theme_bw() +
  theme(axis.title = element_text(size = 12),
        axis.text = element_text(size = 10))

a2 <- fcounts_Fmat_ggplot %>% ggplot(aes(y=values, x = samples, col = group)) + 
  geom_boxplot() +
    ggtitle("counts (log2) (before rlog normalisation)") +
#  scale_color_manual(values = redblue_col) +
  ylab("counts (log2)") +
  theme_bw() +
  theme(axis.title = element_text(size = 12),
        axis.title.x = element_blank(),
        axis.text = element_text(size = 10),
        axis.text.x = element_text(size = 5, angle = 45, hjust = 1))
```

```{r fig.width=12}
plot_grid(a1,a2)
```

# filtered and normalised for lib size

```{r echo=FALSE, results=FALSE}
fcounts_normmat_ggplot <- norm_mat %>% t %>% data.frame
fcounts_normmat_ggplot$samples <- rownames(fcounts_normmat_ggplot)

fcounts_normmat_ggplot$samples == sample_information$name
fcounts_normmat_ggplot$group <- sample_information$treatment

fcounts_normmat_ggplot <- tidyr::gather(fcounts_normmat_ggplot, key = genes, value = values, -samples, -group)

fcounts_normmat_ggplot$samples <- factor(fcounts_normmat_ggplot$samples, levels = unique(fcounts_normmat_ggplot$samples))

# filtered for genes that are not expressed or lowly expressed in all 10 samples

a1 <- fcounts_normmat_ggplot %>% ggplot(aes(x=values, group = samples, col = group)) + 
  geom_density() +
  ggtitle("counts (log2) (lib size normalised)") +
#  scale_color_manual(values = redblue_col) +
  xlab("counts (log2)") +
  theme_bw() +
  theme(axis.title = element_text(size = 12),
        axis.text = element_text(size = 10))

a2 <- fcounts_normmat_ggplot %>% ggplot(aes(y=values, x = samples, col = group)) + 
  geom_boxplot() +
    ggtitle("counts (log2) (lib size normalised)") +
#  scale_color_manual(values = redblue_col) +
  ylab("counts (log2)") +
  theme_bw() +
  theme(axis.title = element_text(size = 12),
        axis.title.x = element_blank(),
        axis.text = element_text(size = 10),
        axis.text.x = element_text(size = 5, angle = 45, hjust = 1))
```

```{r fig.width=12}
plot_grid(a1,a2)
```


## rlog norm 
```{r echo=FALSE, results=FALSE}
rld.ggplot <- rld %>% t %>% data.frame
rld.ggplot$samples <- rownames(rld.ggplot)

rld.ggplot$samples == sample_information$name
rld.ggplot$group <- sample_information$treatment

rld.ggplot <- tidyr::gather(rld.ggplot, key = genes, value = values, -samples, -group)

rld.ggplot$samples <- factor(rld.ggplot$samples, levels = unique(rld.ggplot$samples))

a1 <- rld.ggplot %>% ggplot(aes(x=values, group = samples, col = group)) + 
  geom_density() +
#  scale_color_manual(values = redblue_col) +
  xlab("rlog") +
  ggtitle("rlog normalised") +
  theme_bw() +
  theme(axis.title = element_text(size = 12),
        axis.text = element_text(size = 10))

a2 <- rld.ggplot %>% ggplot(aes(y=values, x = samples, col = group)) + 
  geom_boxplot() +
#  scale_color_manual(values = redblue_col) +
  ylab("normalised expression (rlog)") +
  ggtitle("rlog normalised") +
  theme_bw() +
  theme(axis.title = element_text(size = 12),
        axis.title.x = element_blank(),
        axis.text = element_text(size = 10),
        axis.text.x = element_text(size = 5, angle = 45, hjust = 1))
```

```{r fig.width=12}
plot_grid(a1,a2)
```

# PCA plot

```{r echo=FALSE, results=FALSE}
# Generating the PC and lncRNA matrix (separating them out)
t2g_aug2020 %>% dim
t2g_aug2020 %>% head
t2g_aug2020$gene_biotype %>% table

# out of the filtered genes, 17.5% are lncRNA and 72% are protein-coding and 4% are processed_seudogene
t2g_aug2020$gene_biotype[match(rownames(rld), t2g_aug2020$ext_gene)] %>% table %>% prop.table %>% sort(decreasing = T)
  
# "non_coding", "3prime_overlapping_ncRNA","antisense", "bidirectional_promoter_lncRNA", "lincRNA" ,"macro_lncRNA", "sense_intronic", "sense_overlapping"
index <- rownames(rld) %in% dplyr::filter(t2g_aug2020,gene_biotype %in% c("lncRNA"))$ext_gene
rld_lncRNA <- rld[index,]
# 4676 lncRNAs
dim(rld_lncRNA)

index <- rownames(rld) %in% dplyr::filter(t2g_aug2020,gene_biotype  == "protein_coding")$ext_gene
rld_pc <- rld[index,]
# 14013 protein-coding genes
dim(rld_pc)
```


## PCA plot - protein-coding

Preparing the data frame
```{r echo=FALSE, results=FALSE}
ntop = 1000
Pvars <- rowVars(rld_pc)
select <- order(Pvars, decreasing = TRUE)[1:ntop]

df.PCA_pc_1000 <- prcomp(t(rld_pc[select,]), center =TRUE, scale = TRUE)$x %>% data.frame
# df.PCA_pc_all <- prcomp(t(rld_pc), center =TRUE, scale = TRUE)$x %>% data.frame

prcomp(t(rld_pc[select,]), center =TRUE, scale = TRUE) %>% summary

pca_results <- prcomp(t(rld_pc[select,]), center =TRUE, scale = TRUE)
pca_results$rotation[,"PC1"] %>% sort(decreasing=TRUE) %>%  head(n=50)

df.PCA_pc_1000$samples <- df.PCA_pc_1000 %>% rownames
sample_information$samples <- sample_information$name
df.PCA_pc_1000 <- merge(df.PCA_pc_1000, sample_information, by = "samples")
```

```{r}
# the top 50 genes that drives PC1. An arbritrary threshold needs to be applied to get the top genes. 
pca_results$rotation[,"PC1"] %>% sort(decreasing=TRUE) %>%  head(n=100)
pca_results$rotation[,"PC2"] %>% sort(decreasing=TRUE) %>%  head(n=100)
```




```{r eval=FALSE, echo=FALSE}
# test 
install.packages("ggbioplot")
library(ggbioplot)

pca_results
ggbiplot(pca_results)

biplot(pca_results, showLoadings = TRUE,
    labSize = 5, pointSize = 5, sizeLoadingsNames = 5)

ggbiplot(pca_results, obs.scale = 1, var.scale = 1, ellipse = TRUE, circle = TRUE) +
  scale_color_discrete(name = '') +
  theme(legend.direction = 'horizontal', legend.position = 'top')


```


```{r echo=FALSE}
library('PCAtools')

metadata <- sample_information
rownames(metadata) <- metadata$name
pca_results_test <- pca(rld_pc[select,], metadata = metadata, removeVar = 0.1)


screeplot(pca_results_test, axisLabSize = 18, titleLabSize = 22)

biplot(pca_results_test)
options(ggrepel.max.overlaps = Inf)
biplot(pca_results_test, showLoadings = TRUE,
    labSize = 2, pointSize = 5, sizeLoadingsNames = 2, ntopLoadings = 10)
```



```{r echo=FALSE, eval=FALSE}
# i need to put this in my blog (eventually ...)

# getting what genes drive the PC's
# Aprils codes + additional info at https://www.biostars.org/p/289196/

# the top 50 genes that drives PC1. An arbritrary threshold needs to be applied to get the top genes. 
pca_results$rotation[,"PC1"] %>% sort(decreasing=TRUE) %>%  head(n=50)

# write.table(pca_results$rotation, file = "/researchers/antonio.ahn/Goel_Labmembers/AprilWatt/2.LY_Fulv_Combo_ERpaper/2.LY_Fulv_Combo_ERpaper/1.ZR-75-1-D/RNAseq_060921/R_analysis/data/csv_files/ZR751D_LY_Fulv_Comb_PCAloadings.txt", sep = "\t", 
            col.names = NA)
```

```{r echo=FALSE, results=FALSE}
col_npg3 <- c(ggsci::pal_npg()(10)[c(3,4,9)])
col_npg5 <- c(ggsci::pal_npg()(10)[c(3,4,9,1,2)])
col_Dark2 <- brewer.pal(8,"Dark2")

# hjust = 1, vjust = 1.5,
a1 <- dplyr::filter(df.PCA_pc_1000) %>% 
ggplot(aes(x = PC1, y = PC2, col = treatment, shape = knockout)) +
  geom_point(size = 7) + 
  geom_text_repel(aes(label = samples), size = 3) +
#  scale_x_continuous(limits = c(-150, 150)) +
  scale_color_manual(values = col_npg5) +
  labs(shape = "sample", col = "treatment", 
       x = paste0("PC1","(",round(pca_results$sdev[1]^2/sum(pca_results$sdev^2)*100,1), "%)"), 
       y = paste0("PC2","(",round(pca_results$sdev[2]^2/sum(pca_results$sdev^2)*100,1), "%)")) +
  ggtitle("protein coding (top 1000 highest variance)") +
#  facet_grid( . ~ transformation) +
  theme_bw() +
  theme(axis.text = element_text(color = "black", size = 15), axis.title = element_text(color = "black", size = 20), plot.title = element_text(hjust=0.5))

a2 <- dplyr::filter(df.PCA_pc_1000) %>% 
ggplot(aes(x = PC1, y = PC3, col = treatment, shape = knockout)) +
  geom_point(size = 7) + 
  geom_text_repel(aes(label = samples), size = 3) +
#  scale_x_continuous(limits = c(-150, 150)) +
  scale_color_manual(values = col_npg5) +
  labs(shape = "sample", col = "treatment", x = paste0("PC1","(",round(pca_results$sdev[1]^2/sum(pca_results$sdev^2)*100,1), "%)"), 
       y = paste0("PC3","(",round(pca_results$sdev[3]^2/sum(pca_results$sdev^2)*100,1), "%)")) +
  ggtitle("protein coding (top 1000 highest variance)") +
#  facet_grid( . ~ transformation) +
  theme_bw() +
  theme(axis.text = element_text(color = "black", size = 15), axis.title = element_text(color = "black", size = 20), plot.title = element_text(hjust=0.5))

a3 <- dplyr::filter(df.PCA_pc_1000) %>% 
ggplot(aes(x = PC2, y = PC3, col = treatment, shape = knockout)) +
  geom_point(size = 7) + 
  geom_text_repel(aes(label = samples), size = 3) +
#  scale_x_continuous(limits = c(-150, 150)) +
  scale_color_manual(values = col_npg5) +
  labs(shape = "sample", col = "treatment", 
       x = paste0("PC2","(",round(pca_results$sdev[2]^2/sum(pca_results$sdev^2)*100,1), "%)"), 
       y = paste0("PC3","(",round(pca_results$sdev[3]^2/sum(pca_results$sdev^2)*100,1), "%)")) +
  ggtitle("protein coding (top 1000 highest variance)") +
#  facet_grid( . ~ transformation) +
  theme_bw() +
  theme(axis.text = element_text(color = "black", size = 15), axis.title = element_text(color = "black", size = 20), plot.title = element_text(hjust=0.5))


a4 <- dplyr::filter(df.PCA_pc_1000) %>% 
ggplot(aes(x = PC1, y = PC4, col = treatment, shape = knockout)) +
  geom_point(size = 7) + 
  geom_text_repel(aes(label = samples), size = 3) +
#  scale_x_continuous(limits = c(-150, 150)) +
  scale_color_manual(values = col_npg5) +
  labs(shape = "sample", col = "treatment", 
       x = paste0("PC1","(",round(pca_results$sdev[1]^2/sum(pca_results$sdev^2)*100,1), "%)"), 
       y = paste0("PC4","(",round(pca_results$sdev[4]^2/sum(pca_results$sdev^2)*100,1), "%)")) +
  ggtitle("protein coding (top 1000 highest variance)") +
#  facet_grid( . ~ transformation) +
  theme_bw() +
  theme(axis.text = element_text(color = "black", size = 15), axis.title = element_text(color = "black", size = 20), plot.title = element_text(hjust=0.5))

```

```{r echo=FALSE, fig.height = 8, fig.with = 12, dpi=300}
a1
a2
a3
```

```{r eval=FALSE, echo=FALSE}
setwd("/researchers/krutika.ambani/Goel_lab_members/Keefe_Chan/220714_sgRB1_sgTP53_sgControl_DMSO_DOX_LY/R_analysis/220714_intiial_QC_AA/1.QC_exploratory/figures_output")
png("PCA_pc_plot1.png",        
    width = 6*400,        # 5 x 300 pixels
    height = 6*400,      # 5 x 400 pixels
    res = 250)            # 300 pixels per inch
#    pointsize = 10)        # smaller font size


a1

dev.off()
```

### PC3 protein-coding genes:  further analysis


```{r echo=FALSE, results=FALSE}
# Single gene plots
norm_mat_all_ggplot <- readRDS(file = "/researchers/krutika.ambani/Goel_lab_members/Keefe_Chan/220714_sgRB1_sgTP53_sgControl_DMSO_DOX_LY/R_analysis/220714_intiial_QC_AA/2.Differential_expression/data/rds_files/norm_mat_ggplot_allsamples.rds")

norm_mat_all_ggplot$name  %>% tail


norm_mat_all_ggplot$treatment_group %>% factor %>% levels
#norm_mat_all_ggplot$treatment_group <- factor(norm_mat_all_ggplot$treatment_group,
#                                              levels = c("DMSO_sgControl_1", "LY_sgControl_1", "DOX_sgControl_1",
#                                                         "DMSO_sgControl_3", "LY_sgControl_3", "DOX_sgControl_3",
#                                                         "DMSO_sgTP53_1.3", "LY_sgTP53_1.3", "DOX_sgTP53_1.3",
#                                                         "DMSO_sgTP53_4.2", "LY_sgTP53_4.2", "DOX_sgTP53_4.2",
#                                                         "DMSO_sgRB1_2.6", "DOX_sgRB1_2.6", 
#                                                         "DMSO_sgRB1_4.1", "DOX_sgRB1_4.1"))

norm_mat_all_ggplot$treatment_group <- factor(norm_mat_all_ggplot$treatment_group,
                                              levels = c("DMSO_sgControl_1", "LY_sgControl_1", "DOX_sgControl_1",
                                                         "DMSO_sgControl_3", "LY_sgControl_3", "DOX_sgControl_3",
                                                         "DMSO_sgTP53_1-3", "LY_sgTP53_1-3", "DOX_sgTP53_1-3",
                                                         "DMSO_sgTP53_4-2", "LY_sgTP53_4-2", "DOX_sgTP53_4-2",
                                                         "DMSO_sgRB1_2-6", "DOX_sgRB1_2-6", 
                                                         "DMSO_sgRB1_4-1", "DOX_sgRB1_4-1"))



# add another column called knockout_clone to separate out the groups in the plot
norm_mat_all_ggplot <- norm_mat_all_ggplot %>% unite(col = "knockout_clone", knockout, clone, remove=FALSE) 
# order the group
norm_mat_all_ggplot$knockout_clone <- norm_mat_all_ggplot$knockout_clone %>% 
  factor(levels = c("sgControl_1", "sgControl_3" , 
                    "sgTP53_1-3", "sgTP53_4-2", 
                    "sgRB1_2-6", "sgRB1_4-1"))

ggplot_gene_all <- function(gene){
dplyr::filter(norm_mat_all_ggplot, genesym == gene) %>% 
  ggplot(aes(y = values, x = treatment_group)) + 
    facet_grid( . ~ knockout_clone, scales = "free") + 
#  geom_boxplot(outlier.shape = NA) + 
  geom_point(aes(col = treatment), alpha = 1, size = 5) +
#  geom_line(aes(group=lines), col="red") + 
# geom_jitter(alpha = 1, size = 2, position = position_jitter(seed = 1)) + 
    scale_color_manual(values = col_npg5) + 
    ggtitle(gene) +
    #geom_text(aes(label = name), size = 1.5, vjust = 1) +
    geom_text_repel(aes(label = replicate), size = 1.5, vjust = 1) +
    theme_bw() + 
    scale_y_continuous("gene expression (log2)") +
    theme(axis.text.x =element_text(size=10, angle = 45, hjust=1),
       axis.title.x.bottom=element_blank(),
       axis.title.y=element_text(size=25), legend.position = "none", plot.title = element_text(hjust=0.5,size=20))
}
```


```{r}
ggplot_gene_all("KRT20")
ggplot_gene_all("KRT4")
ggplot_gene_all("EMP1")
ggplot_gene_all("UGT1A6")
ggplot_gene_all("ABCA12")
ggplot_gene_all("DIO2")
ggplot_gene_all("ASB9")

ggplot_gene_all("IL6")
ggplot_gene_all("MMP1")
ggplot_gene_all("IL32")
ggplot_gene_all("PLAUR")

```

```{r}
PC3_100genes <- pca_results$rotation[,"PC3"] %>% sort(decreasing=TRUE) %>%  head(n=100) %>% names

heatmap_mat <- norm_mat[PC3_100genes,]
# generate z-score
scaled.mat <- t(scale(t(heatmap_mat))) 

### annotate column
columnfactor <- sample_information %>% 
  unite(col = "group" , c(knockout,clone,treatment), sep = "_") %>% .$group


annotation_column <- data.frame(row.names = colnames(scaled.mat), group = columnfactor)
# 8 groups to color
annotation_column$group %>% unique %>% length

### annotate rows.  no need here
#table(rownames(scaled.mat) == names(cut_output))
#annotation_row <- data.frame(row.names = names(rownames_cluster_new), cluster = #factor(paste0("c",rownames_cluster_new), levels = c(paste0("c",1:20))))

### annotate colors
# annotate colors: column
#startrek_col <- ggsci::pal_startrek()(7)[1:4]
#names(startrek_col) <- c("DMSO","DOX","LY","RAD")

#paired_color <- brewer.pal(8,"Paired")
#names(paired_color) <- unique(annotation_column$group)
#npg_col <- pal_npg()(8)[3:4] 
#names(npg_col) <- c("sig", "nonsig")
# annotate colors: rows
# row_colors <- c(brewer.pal(12,"Paired")[9:10], brewer.pal(8,"Dark2"))
column_colors <- c(brewer.pal(8,"Dark2"), ggsci::pal_simpsons()(8))
names(column_colors) <- unique(annotation_column$group)
#row_colors <-  colorRampPalette(c("blue", "purple","green", "white", "orange","red"))(20)
#names(row_colors) <- paste0("c",1:20)
#row_colors <- pals::stepped()[1:20]
#names(row_colors) <- paste0("c",1:20)

# annotate colors: put all together
anno_colors <- list(group = column_colors)

### other parameters
scaled.mat %>% max
scaled.mat %>% min
breaksList <- seq(-3, 3, length.out = 100)

# col.pan <- colorRampPalette(c("purple","black", "yellow"))(100)
col.pan <-  colorRampPalette(c("black","dodgerblue3", "white", "orange","red"))(100)
```

```{r fig.height=15, fig.width=12}
pheatmap::pheatmap(scaled.mat, col= col.pan, cluster_rows = F, cluster_cols = F, show_rownames = T, show_colnames=T, annotation_col = annotation_column, border_color = FALSE, legend = T,fontsize_row = 4,annotation_colors = anno_colors, treeheight_col = 20, breaks = breaksList, clustering_distance_rows = "euclidean",clustering_method = "complete")
```


```{r eval=FALSE, echo=FALSE}
setwd("/researchers/krutika.ambani/Goel_lab_members/Keefe_Chan/220714_sgRB1_sgTP53_sgControl_DMSO_DOX_LY/R_analysis/220714_intiial_QC_AA/1.QC_exploratory/figures_output")

png("PCA_pc_plot3_heatmap.png",        
    width = 5*400,        # 5 x 300 pixels
    height = 5*400,      # 5 x 400 pixels
    res = 250)            # 300 pixels per inch
#    pointsize = 10)        # smaller font size


pheatmap::pheatmap(scaled.mat, col= col.pan, cluster_rows = F, cluster_cols = F, show_rownames = T, show_colnames=T, annotation_col = annotation_column, border_color = FALSE, legend = T,fontsize_row = 4,annotation_colors = anno_colors, treeheight_col = 20, breaks = breaksList, clustering_distance_rows = "euclidean",clustering_method = "complete")

dev.off()
```

### hypergeometric test for PC3 genes

```{r message=FALSE}
library(msigdbr)
library(clusterProfiler)
h_gene_sets = msigdbr(species = "Homo sapiens", category = "H")

c2_gene_sets = msigdbr(species = "Homo sapiens", category = "C2")
c2_gene_sets <- c2_gene_sets %>% filter(grepl("REACTOME|KEGG", c2_gene_sets$gs_name))

c5_gene_sets = msigdbr(species = "Homo sapiens", category = "C5")
c5_gene_sets <- c5_gene_sets %>% filter(!grepl("HP_", c5_gene_sets$gs_name))

MSigDB.ofinterest <- rbind(h_gene_sets, c2_gene_sets, c5_gene_sets)
MSigDB.ofinterest <- MSigDB.ofinterest  %>% dplyr::select(gs_name, gene_symbol)

PC3_hypergeo_output <-   enricher(PC3_100genes, TERM2GENE = MSigDB.ofinterest)
```

```{r}
PC3_hypergeo_output %>% as_tibble %>% DT::datatable()
```



## PCA plot - lncRNA 

```{r echo=FALSE, results=FALSE}
ntop = 500
Pvars <- rowVars(rld_lncRNA)
select <- order(Pvars, decreasing = TRUE)[1:ntop]

df.PCA_lncRNA_500 <- prcomp(t(rld_lncRNA[select,]), center =TRUE, scale = TRUE)$x %>% data.frame
# df.PCA_lncRNA_all <- prcomp(t(rld_lncRNA), center =TRUE, scale = TRUE)$x %>% data.frame

pca_lncRNA_results <- prcomp(t(rld_lncRNA[select,]), center =TRUE, scale = TRUE)

prcomp(t(rld_lncRNA[select,]), center =TRUE, scale = TRUE) %>% summary

df.PCA_lncRNA_500$samples <- df.PCA_lncRNA_500 %>% rownames
df.PCA_lncRNA_500 <- merge(df.PCA_lncRNA_500, sample_information, by = "samples")

# df.PCA_lncRNA_all$samples <- df.PCA_lncRNA_all %>% rownames
# df.PCA_lncRNA_all <- merge(df.PCA_lncRNA_all, sample_information, by = "samples")
```

```{r}
pca_results_lncRNA <- prcomp(t(rld_lncRNA[select,]), center =TRUE, scale = TRUE) 
# the top 50 genes that drives PC1. An arbritrary threshold needs to be applied to get the top genes. 
pca_results_lncRNA$rotation[,"PC1"] %>% sort(decreasing=TRUE) %>%  head(n=100)
pca_results_lncRNA$rotation[,"PC2"] %>% sort(decreasing=TRUE) %>%  head(n=100)
```

```{r}
a1 <- dplyr::filter(df.PCA_lncRNA_500) %>% 
ggplot(aes(x = PC1, y = PC2, col = treatment, shape = knockout)) +
  geom_point(size = 7) + 
  geom_text_repel(aes(label = samples), size = 3) +
#  scale_x_continuous(limits = c(-150, 150)) +
  scale_color_manual(values = col_npg5) +
  labs(shape = "sample", col = "treatment", 
       x = paste0("PC1","(",round(pca_lncRNA_results$sdev[1]^2/sum(pca_lncRNA_results$sdev^2)*100,1), "%)"), 
       y = paste0("PC2","(",round(pca_lncRNA_results$sdev[2]^2/sum(pca_lncRNA_results$sdev^2)*100,1), "%)")) +
  ggtitle("lncRNA (top 500 highest variance)") +
#  facet_grid( . ~ transformation) +
  theme_bw() +
  theme(axis.text = element_text(color = "black", size = 15), axis.title = element_text(color = "black", size = 20), plot.title = element_text(hjust=0.5))

a2 <- dplyr::filter(df.PCA_lncRNA_500) %>% 
ggplot(aes(x = PC1, y = PC3, col = treatment, shape = knockout)) +
  geom_point(size = 7) + 
  geom_text_repel(aes(label = samples), size = 3) +
#  scale_x_continuous(limits = c(-150, 150)) +
  scale_color_manual(values = col_npg5) +
  labs(shape = "sample", col = "treatment", x = paste0("PC1","(",round(pca_lncRNA_results$sdev[1]^2/sum(pca_lncRNA_results$sdev^2)*100,1), "%)"), 
       y = paste0("PC3","(",round(pca_lncRNA_results$sdev[3]^2/sum(pca_lncRNA_results$sdev^2)*100,1), "%)")) +
  ggtitle("lncRNA (top 500 highest variance)") +
#  facet_grid( . ~ transformation) +
  theme_bw() +
  theme(axis.text = element_text(color = "black", size = 15), axis.title = element_text(color = "black", size = 20), plot.title = element_text(hjust=0.5))

a3 <- dplyr::filter(df.PCA_lncRNA_500) %>% 
ggplot(aes(x = PC2, y = PC3, col = treatment, shape = knockout)) +
  geom_point(size = 7) + 
  geom_text_repel(aes(label = samples), size = 3) +
#  scale_x_continuous(limits = c(-150, 150)) +
  scale_color_manual(values = col_npg5) +
  labs(shape = "sample", col = "treatment", 
       x = paste0("PC2","(",round(pca_lncRNA_results$sdev[2]^2/sum(pca_lncRNA_results$sdev^2)*100,1), "%)"), 
       y = paste0("PC3","(",round(pca_lncRNA_results$sdev[3]^2/sum(pca_lncRNA_results$sdev^2)*100,1), "%)")) +
  ggtitle("lncRNA (top 500 highest variance)") +
#  facet_grid( . ~ transformation) +
  theme_bw() +
  theme(axis.text = element_text(color = "black", size = 15), axis.title = element_text(color = "black", size = 20), plot.title = element_text(hjust=0.5))

```

```{r echo=FALSE, fig.height = 8, fig.with = 12, dpi=300}
a1
a2
a3
```



# hierarchical cluster

```{r echo=FALSE, results=FALSE}
d <- dist(t(rld_pc)) 
hc <- hclust(d)
dend.pc =as.dendrogram(hc)


ddata.pc <- dendro_data(dend.pc, type = "rectangle")
clust <- cutree(hc ,k=16)  
clust.df <- data.frame(label=names(clust), 
                        group = factor(clust))
# levels(clust.df$group) <- c("inducible", "constitutive")

ddata.pc$labels <- merge(ddata.pc$labels,clust.df, by="label")


a1 <- ggplot() + geom_segment(data = segment(ddata.pc), aes(x = x, y = y, xend = xend, yend = yend)) + 
  coord_flip() + 
  scale_y_reverse(expand = c(0.3, 0)) + 
  geom_text(data = ddata.pc$labels, aes(x, y, label=label, hjust=0, col = group), size=4) + 
  ggtitle("mRNA") +
#  scale_color_manual(values = col_npg5) + 
#  theme_dendro() + 
  theme_bw() +
  theme(legend.position="none", axis.title = element_blank(), axis.text = element_text(size = 12), plot.title = element_text(hjust=0.5, size = 20))
```

```{r fig.height = 12, fig.width=12}
a1
```

```{r echo=FALSE, results=FALSE}
d <- dist(t(rld_lncRNA)) 
hc <- hclust(d)
dend.lncRNA =as.dendrogram(hc)


ddata.lncRNA <- dendro_data(dend.lncRNA, type = "rectangle")
clust <- cutree(hc ,k=16)  
clust.df <- data.frame(label=names(clust), 
                        group = factor(clust))
# levels(clust.df$group) <- c("inducible", "constitutive")

ddata.lncRNA$labels <- merge(ddata.lncRNA$labels,clust.df, by="label")
 
a2 <- ggplot() + geom_segment(data = segment(ddata.lncRNA), aes(x = x, y = y, xend = xend, yend = yend)) + 
  coord_flip() + 
  scale_y_reverse(expand = c(0.3, 0)) + 
  geom_text(data = ddata.lncRNA$labels, aes(x, y, label=label, hjust=0, col = group), size=4) + 
  ggtitle("lncRNA") +
#  scale_color_manual(values = col_npg5) + 
#  theme_dendro() + 
  theme_bw() +
  theme(legend.position="none", axis.title = element_blank(), axis.text = element_text(size = 12), plot.title = element_text(hjust=0.5, size = 20))
```

```{r fig.height = 12, fig.width=12}
a2
```

# save count matrix


```{r eval=FALSE, echo=FALSE}
write.csv(fcounts_mat_genesymbol, file = "/researchers/krutika.ambani/Goel_lab_members/Keefe_Chan/220714_sgRB1_sgTP53_sgControl_DMSO_DOX_LY/R_analysis/220714_intiial_QC_AA/1.QC_exploratory/data/csv_files/MCF7M_sgRB1_sgTP53_sgControl_rawcounts_mat.csv")

write.csv(norm_mat, file = "/researchers/krutika.ambani/Goel_lab_members/Keefe_Chan/220714_sgRB1_sgTP53_sgControl_DMSO_DOX_LY/R_analysis/220714_intiial_QC_AA/1.QC_exploratory/data/csv_files/MCF7M_sgRB1_sgTP53_sgControl_normalised_mat.csv")

write.csv(sample_information, file = "/researchers/krutika.ambani/Goel_lab_members/Keefe_Chan/220714_sgRB1_sgTP53_sgControl_DMSO_DOX_LY/R_analysis/220714_intiial_QC_AA/1.QC_exploratory/data/csv_files/MCF7M_sgRB1_sgTP53_sgControl_TIS_sample_information.csv")

write.csv(fcounts_mat, file = "/researchers/krutika.ambani/Goel_lab_members/Keefe_Chan/220714_sgRB1_sgTP53_sgControl_DMSO_DOX_LY/R_analysis/220714_intiial_QC_AA/1.QC_exploratory/data/csv_files/MCF7M_sgRB1_sgTP53_sgControl_TIS_fcounts_mat.csv" )
```



# save RData

```{r eval=FALSE}
save.image("/researchers/krutika.ambani/Goel_lab_members/Keefe_Chan/220714_sgRB1_sgTP53_sgControl_DMSO_DOX_LY/R_analysis/220714_intiial_QC_AA/1.QC_exploratory/data/QC_exploratory.RData")

load("/researchers/krutika.ambani/Goel_lab_members/Keefe_Chan/220714_sgRB1_sgTP53_sgControl_DMSO_DOX_LY/R_analysis/220714_intiial_QC_AA/1.QC_exploratory/data/QC_exploratory.RData")
```

